<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Treefit - Working with Seurat</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Treefit - Working with Seurat</h1>



<p>This is a user-friendly manual for biologists who are interested in
performing quantitative trajectory inference with the Treefit package.
If you are not familiar with the basic workflow of Treefit, then it
would be a good idea to look at the introductory tutorial
(<code>vignette(&quot;treefit&quot;)</code>) before using it.</p>
<div id="why-quantitative-trajectory-inference" class="section level2">
<h2>1. Why “quantitative” trajectory inference?</h2>
<p>Single-cell technologies are expected to help us discover a novel
type of cells and revolutionize our understanding of the process of cell
differentiation, and trajectory inference is one of the key
computational challenges in single-cell transcriptomics. In recent
years, many software packages have been developed and widely used to
extract an underlying “tree” trajectory from single-cell RNA-seq
data.</p>
<p>However, trajectory inference often suffers from the uncertainty due
to the heterogeneity of individual cells and the high levels of
technical noise in single-cell experiments. Hence, it was a fatal
problem that there was no method to quantitatively assess the
reliability of the estimated trajectory and to distinguish distinct cell
types in an objective manner so far. Of course, many handy software
packages are available for visualizing data and helping perform
exploratory data analysis; however, we must stress that different
visualization techniques often result in completely different pictures
and the interpretation can be quite subjective and thus it is hard to
reach a scientific truth solely by this approach. In order to facilitate
scientific discoveries from single-cell gene expression data, we need a
<em>quantitative</em> methodology.</p>
<p>This is why we developed Treefit - the first software for performing
<em>quantitative</em> trajectory inference. Treefit can be used in
conjunction with existing popular software packages such as <a href="https://satijalab.org/seurat/">Seurat</a> and <a href="https://dynverse.org/">dynverse</a>. In this tutorial, we
demonstrate how to use Treefit together with Seurat by analyzing the
single-cell RNA-seq dataset in <a href="https://www.nature.com/articles/nbt.2859">Trapnell <i>el al.</i>,
2014, Nature Biotechnology</a>.</p>
</div>
<div id="trajectory-inference-using-treefit-and-seurat" class="section level2">
<h2>2. Trajectory inference using Treefit and Seurat</h2>
<p>As the Treefit package is meant to be used with gene expression data
(either raw counts or normalized values) that have been already quality
controlled, it is a good idea to use Seurat or other popular toolkits
designed for the quality control of single-cell gene expression data
(<i>e.g.</i>, filtering cells and normalizing the data) before the
Treefit analysis.</p>
<div id="loading-the-dataset" class="section level3">
<h3>2.1. Loading the dataset</h3>
<p>We analyze the dataset provided in <a href="https://www.nature.com/articles/nbt.2859">Trapnell <i>el al.</i>,
2014, Nature Biotechnology</a>. This dataset was originally acquired for
the purpose of studying the process of differentiation of human
myoblasts that can be reasonably explained by the linear model
(<i>i.e.</i>, non-branching, chain-like structure). However, as the
authors discuss in detail in the original article, their trajectory
inference software based on a minimum spanning tree (MST) helped detect
contamination with another cell lineage which was not related to
myogenesis and so revealed that the bifurcation tree model (<i>i.e.</i>,
Y-shaped tree structure) fits the data better than the linear model.</p>
<p>To use this dataset, we download and read the dataset provided in
dynverse as follows (see <a href="https://zenodo.org/record/1443566" class="uri">https://zenodo.org/record/1443566</a> for details).</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>trapnell.path <span class="ot">&lt;-</span> <span class="st">&quot;myoblast-differentiation_trapnell.rds&quot;</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(trapnell.path)) {</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>  <span class="fu">download.file</span>(<span class="st">&quot;https://zenodo.org/record/1443566/files/real/gold/myoblast-differentiation_trapnell.rds?download=1&quot;</span>,</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>                trapnell.path,</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>                <span class="at">mode=</span><span class="st">&quot;wb&quot;</span>)</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>}</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>trapnell.dynverse <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(trapnell.path)</span></code></pre></div>
</div>
<div id="preprocessing-with-seurat" class="section level3">
<h3>2.2. Preprocessing with Seurat</h3>
<p>From now, we demonstrate how to preprocess the above data with
Seurat. In order to handle data in Seurat, we first need to create a
<code>Seurat</code> object. When creating a <code>Seurat</code> object
from dynverse data, we must be aware that the roles of the rows and
columns are reversed and therefore we need to transpose the matrix. In
dynverse data, the rows and columns correspond to cells and genes,
respectively, and vice versa in <code>Seurat</code> objects.</p>
<p>Here we assume that we wish to perform the following
preprocessing.</p>
<ul>
<li>Excluding each gene that is only expressed in less than 3 cells</li>
<li>Removing each cell that only expresses less than 200 genes</li>
</ul>
<p>This can be done by setting the designated parameters as follows.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>trapnell <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">CreateSeuratObject</span>(<span class="at">counts=</span><span class="fu">t</span>(trapnell.dynverse<span class="sc">$</span>count),</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>                                       <span class="at">min.cells=</span><span class="dv">3</span>,</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>                                       <span class="at">min.features=</span><span class="dv">200</span>)</span></code></pre></div>
<pre><code>## Warning: Feature names cannot have underscores (&#39;_&#39;), replacing with dashes
## (&#39;-&#39;)</code></pre>
<pre><code>## Warning: Data is of class matrix. Coercing to dgCMatrix.</code></pre>
<p>The “Feature names cannot have underscores …” warning message can be
ignored. It’s caused by difference between dynverse and Seurat. Dynverse
uses underscores for gene names but Seurat doesn’t allow it. Seurat
replaces underscores with dashes automatically with the warning
message.</p>
</div>
<div id="visualization-with-seurat" class="section level3">
<h3>2.3. Visualization with Seurat</h3>
<p>Although pictures may not always describe reality, we usually
visualize data before starting detailed analysis. Here we create a
scatter plot by principal component analysis (PCA). Below are a few
lines of code to do this and the result is shown in Figure 1.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>trapnell <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">FindVariableFeatures</span>(trapnell, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>trapnell <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">ScaleData</span>(trapnell, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## Warning: No layers found matching search pattern provided</code></pre>
<pre><code>## Error in `ScaleData()`:
## ! No layer matching pattern &#39;data&#39; found. Please run NormalizeData and retry</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>trapnell <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">RunPCA</span>(trapnell, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## Warning: No layers found matching search pattern provided</code></pre>
<pre><code>## Error in `PrepDR5()`:
## ! No layer matching pattern &#39;scale.data&#39; not found. Please run ScaleData and retry</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">plot</span>(Seurat<span class="sc">::</span><span class="fu">Embeddings</span>(trapnell))</span></code></pre></div>
<pre><code>## Error in `object[[reduction]]`:
## ! &#39;pca&#39; not found in this Seurat object
## </code></pre>
<p>It is hard to reach the truth from the PCA result shown in Figure 1.
Unlike the tree-like toy data analyzed in the previous tutorial
(<code>vignette(&quot;treefit&quot;)</code>), we cannot easily recognize the
underlying tree structure for various reasons (<i>e.g.</i>, high level
of noise, contamination with different types of cells, and the
inaccuracy of two-dimensional visualization of high dimensional
data).</p>
<p>Figure 2b in <a href="https://www.nature.com/articles/nbt.2859">Trapnell <i>el al.</i>,
2014, Nature Biotechnology</a> shows that this dataset contains three
distinct types of cells (<i>i.e.</i>, proliferating cells,
differentiating myoblasts, and contaminating interstitial mesenchymal
cells), which led the authors to adopt a Y-shaped tree model
(<i>i.e.</i>, a star tree with three arms) to explain the data. However,
it is difficult to quantify the reliability of this tree model and to
confidently predict that the best-fit tree consists of three principal
paths.</p>
</div>
</div>
<div id="quantitative-trajectory-inference-with-treefit" class="section level2">
<h2>3. Quantitative trajectory inference with Treefit</h2>
<div id="performing-the-treefit-analysis" class="section level3">
<h3>3.1. Performing the Treefit analysis</h3>
<p>Let us use Treefit in order to estimate the tree-likeness of the
preprocessed myoblast data (which contain 290 cells) and to predict the
number of principal paths in the best-fit tree. The workflow is the same
as the toy data analysis described in the previous tutorial
(<code>vignette(&quot;treefit&quot;)</code>).</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>trapnell.fit <span class="ot">&lt;-</span> treefit<span class="sc">::</span><span class="fu">treefit</span>(trapnell)</span></code></pre></div>
<pre><code>## Error in `Seurat::GetAssayData()`:
## ! `assay` must be one of &quot;RNA&quot;, not &quot;counts&quot;.</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>trapnell.fit</span></code></pre></div>
<pre><code>## Error: オブジェクト &#39;trapnell.fit&#39; がありません</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="fu">plot</span>(trapnell.fit)</span></code></pre></div>
<pre><code>## Error: オブジェクト &#39;trapnell.fit&#39; がありません</code></pre>
</div>
<div id="interpreting-the-results" class="section level3">
<h3>3.2. Interpreting the results</h3>
<p>As explained in the previous tutorial
(<code>vignette(&quot;treefit&quot;)</code>), the two Grassmann distances
<code>$max_cca_distance</code> and <code>$rms_cca_distance</code> play
an essential role in the Treefit analysis. The goal of the first
analysis using <code>$max_cca_distance</code> is to measure the
goodness-of-fit between data and data-derived tree trajectories, and
this helps biologists make a mathematical and statistical evidence-based
quantitative discussion. The other analysis with
<code>$rms_cca_distance</code> aims to predict the number of principal
paths in the best-fit tree trajectory, and the result of this analysis
can be helpful to discover novel cell types or make sure whether or not
there are contaminating cell types.</p>
<p>If we compare the mean values of <code>$max_cca_distance</code> shown
in the left panel of Figure 2 with the previous results
(<code>vignette(&quot;treefit&quot;)</code>), we can see that the tree-likeness of
the myoblast dataset is more like the noisy tree-like data with the
<code>fatness</code> <code>0.8</code> rather than the tree-like data
with the <code>fatness</code> <code>0.1</code>. This estimate is
reasonable in light of the fact that the myoblast dataset in <a href="https://www.nature.com/articles/nbt.2859">Trapnell <i>el al.</i>,
2014, Nature Biotechnology</a> contains some contaminating cell types
that are not related to myogenesis.</p>
<p>Turning to the right panel of Figure 2, we see that
<code>$rms_cca_distance</code> attains the local minimum at <i>p</i>=2.
Therefore, we can deduce that the best-fit tree trajectory for the
myoblast data consists of <i>p</i>+1=3 principal paths.
<code>$n_principal_paths_candidates[1]</code> also tells us the number
principal paths is 3. Recalling the discussion in <a href="https://www.nature.com/articles/nbt.2859">Trapnell <i>el al.</i>,
2014, Nature Biotechnology</a> (<i>i.e.</i>, the cells in the dataset
can be classified into the following three groups: proliferating cells;
differentiating myoblasts; and contaminating interstitial mesenchymal
cells), we thus see that Treefit has correctly predicted the number of
distinct principal paths forming the best-fit tree trajectory for this
dataset.</p>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
